core/include/cubos/core/ecs/dispatcher.hpp

```diff
@@ -9,81 +9,198 @@
+
+        /// Compiles a call chain. This takes all pending systems and
+        /// determines their execution order. This is required before
+        /// systems can be run.
+        void compileChain();
+
+        /// Calls all systems in order of the stages they are in.
+        /// The call chain must be compiled successfully prior, otherwise
+        /// this function won't run anything.
+        /// @param world World to call the systems in.
+        /// @param cmds Command buffer.
+        void callSystems(World& world, CommandBuffer& cmds);
+
+        /// Visits a DFSNode to create a topological order.
+        /// This is used internally during call chain compilation.
+        /// @param node The node to visit.
+        /// @param nodes Array of DFSNodes.
+        /// @return True if a cycle was detected, false if otherwise.
+        bool dfsVisit(DFSNode& node, std::vector<DFSNode>& nodes);
 
     private:
-        std::vector<std::string> stagesOrder; ///< Order in which stages are dispatched.
-        std::map<std::string, std::vector<std::unique_ptr<AnySystemWrapper<void>>>>
-            stagesByName;           ///< Maps names to stages.
-        std::string defaultStage;   ///< The stage that new stages are put after/before.
-        Direction defaultDirection; ///< The direction new stages are put in relation to the default stage.
+        /// Variables for holding information before call chain is compiled.
+        std::vector<System> pendingSystems;                                 ///< All systems.
+        std::map<std::string, std::shared_ptr<SystemSettings>> tagSettings; ///< All tags.
+        System* currSystem;                                                 ///< Last set system, for changing settings.
+        SystemSettings* currTagSettings;                                    ///< Last set tag, for changing settings.
+
+        /// Variables for holding information after call chain is compiled.
+        std::vector<System> systems; ///< Compiled order of running systems.
     };
 
-    template <typename F> void Dispatcher::addSystem(F system, std::string stage)
+    template <typename F> void Dispatcher::addSystem(F func)
+    {
+        // Wrap the system and put it in the pending queue
+        System system = {nullptr, std::make_unique<SystemWrapper<F>>(func)};
+        pendingSystems.push_back(std::move(system));
+        currSystem = &pendingSystems.back();
+    }
+
+    template <typename F> void Dispatcher::systemSetAfterSystem(F func)
     {
-        // Register stage if it doesn't exist.
-        if (this->stagesByName.find(stage) == this->stagesByName.end())
+        auto it = std::find_if(pendingSystems.begin(), pendingSystems.end(), [&func](const System& system) {
+            SystemWrapper<F>* wrapper = dynamic_cast<SystemWrapper<F>*>(system.system.get());
+            if (!wrapper)
+                return false;
+            return wrapper->system == func;
```

Comment:
Can we do this? I didn't know we could compare two different function types like this. Does it work with lambdas?
If it doesn't I'm fine with forcing systems to be actual function pointers.

Reply:
So far it seems to work yeah, although I haven't tested system dependencies extensively. When I get the chance I'll confirm if the current sample works.

As for lambdas I didn't even test them tbh, I'll also try them.

Reply:
Yup, I just checked. It even works with lambdas, which is awesome.

Reply:
I've added system ordering on the example as well to demonstrate.